File: src/5_entities/instances/model/contentType.ts
export const ContentType = {
  Modpack: 'modpack',
  Mod: 'mod',
  DataPack: 'datapack',
  ResourcePack: 'resourcepack',
  ShaderPack: 'shaderpack',
} as const;

export type ContentType = (typeof ContentType)[keyof typeof ContentType];

export type AtomicContentType = Exclude<
  ContentType,
  typeof ContentType.Modpack
>;

export const CONTENT_TYPES = Object.values(
  ContentType,
) as readonly ContentType[];

export const ATOMIC_CONTENT_TYPES: readonly AtomicContentType[] = [
  ContentType.Mod,
  ContentType.DataPack,
  ContentType.ResourcePack,
  ContentType.ShaderPack,
];

export const CONTENT_TYPE_TO_TITLE = {
  [ContentType.Mod]: 'mods',
  [ContentType.DataPack]: 'dataPacks',
  [ContentType.ResourcePack]: 'resourcePacks',
  [ContentType.ShaderPack]: 'shaders',
  [ContentType.Modpack]: 'modpacks',
} as const satisfies Record<ContentType, string>;

export const isContentType = (value: unknown): value is ContentType =>
  typeof value === 'string' && CONTENT_TYPES.includes(value as ContentType);

export const isAtomicContentType = (
  value: unknown,
): value is AtomicContentType =>
  isContentType(value) && value !== ContentType.Modpack;

File: packages/core/aether-core/src/features/plugins/infra/extism/mappers/instance/content_type.rs
use aether_core_plugin_api::v0::ContentTypeDto;

use crate::features::instance::ContentType;

impl From<ContentType> for ContentTypeDto {
    fn from(value: ContentType) -> Self {
        match value {
            ContentType::Modpack => Self::Modpack,
            ContentType::Mod => Self::Mod,
            ContentType::DataPack => Self::DataPack,
            ContentType::ResourcePack => Self::ResourcePack,
            ContentType::ShaderPack => Self::ShaderPack,
        }
    }
}

impl From<ContentTypeDto> for ContentType {
    fn from(value: ContentTypeDto) -> Self {
        match value {
            ContentTypeDto::Modpack => Self::Modpack,
            ContentTypeDto::Mod => Self::Mod,
            ContentTypeDto::DataPack => Self::DataPack,
            ContentTypeDto::ResourcePack => Self::ResourcePack,
            ContentTypeDto::ShaderPack => Self::ShaderPack,
        }
    }
}
